File: C/sortastring.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <ctype.h>

/*
 *  Ok, I admit this one is a bit weird.  It sorts
 *  all the letters in an array by counting the each
 *  of the letters in the array.  I.e., For the word
 *  "mississippi" it would create and array that said
 *  of "i", there are 4.  m is 1, p is 2 and s is 4.
 *  then it build a new array based off of those counts.
 *  Which yields "iiiimppssss".
 *
 *  Weird but it works and it's pretty easy once you see
 *  it and it is quick.
 */

#define ARRLEN 256

void init_array(int *myarray) {
int i;

   for (i = 0; i < ARRLEN; i++) {
      myarray[i] = 0;
   }

   return;
}

void parse_word_string(char *mystring, int *myarray) {
int len, i;

   len = strlen(mystring);
   for (i = 0; i < len; i++) {
      myarray[(unsigned int)tolower(mystring[i])]++;
   }

   return;
}

char *sortedarray(int *myarray) {
char *sorted;
unsigned int i, j, k;

   sorted = (char *)calloc(32, 1);

   k = j = 0;
   for (i = 0; i < ARRLEN; i++) {
      if (myarray[i] > 0) {
         for (j = 0; j < myarray[i]; j++) {
            sorted[k] = (char)i;
            k++;
         }
      }
   }

   return(sorted);
}

void runnit(char *mystring) {
int myarray[ARRLEN];
char *sortedthing;

   init_array(myarray);
   parse_word_string(mystring, myarray);
   sortedthing = sortedarray(myarray);
   printf("original:  %s\n", mystring);
   printf("sorted  :  %s\n", sortedthing);
   free(sortedthing);

}

int main() {

   runnit("hello");
   runnit("massive flood");
   runnit("Mississippi");
   runnit("smile");
   runnit("limes");
   runnit("slime");
   runnit("Multi space with punctuation!");
   runnit("2468097135");
}

File: python/sortastring.py
import sys

#
#   Done for giggles to see
#   how the C sortastring.c
#   would look in python.
#
#   Did it with python lists.
#   You can use actual arrays,
#   buy you'd need to import
#   numpy.
#
def init_array(myarray):
 
   i = 0
   while (i < 256):
      myarray.append(0)
      i += 1

   return(myarray)


def parse_word_string(mystring, myarray):

   #
   #   lowercase everything so the sorts
   #   are consistent with respect to
   #   letter order.
   #
   for i in mystring:
      myarray[ord(i.lower())] += 1

   return(myarray)

def sortedarray(myarray):

   msorted = ""

   for i in range(256):
      if myarray[i] > 0:
         for j in range(myarray[i]):
            msorted = msorted + chr(i)

   return(msorted)

def runnit(mystring):

   myarray = []

   myarray = init_array(myarray)
   myarray = parse_word_string(mystring, myarray)
   msorted = sortedarray(myarray)

   print("original string  : %s" % (mystring))
   print("sorted (program) : %s" % (msorted))
   #
   #   Just so folk realize that I know
   #   python has a number of sorts "built in".
   #
   res = ''.join(sorted(mystring.lower()))
   print("sorted(py sort)  : %s" % (res))

   return


if __name__ == "__main__":

   runnit("Mississippi")
   print("######")
   runnit("massive flood")
   print("######")
   runnit("smile")
   print("######")
   runnit("limes")
   print("######")
   runnit("slime")
   print("######")
   runnit("Multi space with punctuation!")
   print("######")
   runnit("2468097135")
